### DISCUSSION TITLE
Slow queries with payload api, fast with mongoose (x 50 slower?)

### DISCUSSION THREAD

**Discussion by nexflo** [payloadcms/payload#7506] [Q&A]
### Payload Version

3.62

### Node Version

18.17.2

### Next.js Version

latest

### Describe the Bug

We have the following query:
const res = await payload.find({
  collection: "testitems",
  where: { and: [{ listId: { equals: listId } }, { isUsed: { equals: false } }] },
  limit: 1,
});

Via payload api above:
01 find payload api 5734ms
If we use mongoose directly its fast as expected.
02 find mongoose 23ms

Quite critical I believe.

### Reproduction Steps

Write same query in mongoose and payload find api

### Adapters and Plugins

db-mongodb

**Comment by jmikrut** [comment]
Hey @nexflo this could be due to many things but it's definitely something to do with your configuration. 5734ms is definitely not something that should happen with Payload.

If you use Mongoose directly, you're not going through any Payload hooks, access control, or relationship population. So there's lots more that Payload has to do vs. Mongoose. If at any point you have a misconfigured hook, or access control, or slow query from a population / similar, things could slow down significantly. 

You should start by checking if you have any queries that are filtering on fields that you do not have indexed. For example, `listId` or `isUsed`. Do you have those indexed? 

How many documents do you have in the collection that you're querying? Does the speed improve if you add `depth=0` to your `payload.find` query?

I will convert this into a discussion so we can continue to help but I don't think this is related to Payload. If you can give us a reproduction and it does turn out to be something to do with Payload I would be happy to re-open the issue though.

**Reply by nexflo** [reply]
Hmm, no relations, no drafts, all indexed.
We do have 1-10million records, but being close to mongoose native speed...would indeed be essential.

**Reply by nexflo** [reply]
@jmikrut 

Reproduction, i created a collection "testitems" , with all indexed fields:

fields: [
    {
      name: "code",
      type: "text",
      unique: true,
      required: true,
      index: true,
    },
    {
      name: "list",
      label: "List ID",
      type: "text",
      required: true,
      index: true,
    },
    {
      name: "isUsed",
      type: "checkbox",
      defaultValue: false,
      required: true,
      index: true,
    },
  ],

I randomly added 1 millions codes into the DB.

I query the DB as shown above:

const res = await payload.find({
collection: "testitems",
where: { and: [{ listId: { equals: listId } }, { isUsed: { equals: false } }] },
limit: 1,
});

and do the same query with moongose.

Difference is around 5 seconds. (same thing also happens inside the payload admin panel if I add a filter)
I guess payload loops through every query and does checks? Or the query is unoptimised?

Thanks in advance

Flo

**Reply by r1tsuu** [reply]
> Hmm, no relations, no drafts, all indexed. We do have 1-10million records, but being close to mongoose native speed...would indeed be essential.

Could it be actually because of `count`? Like can you call `Model.countDocuments()` directly and see how fast it will be? Because with Postgres I personally had this one issue https://github.com/payloadcms/payload/issues/6321 which leads to a problem with `COUNT`. 
And the reason is simillar - a large amount of records. But 6 seconds still feels too much, I don't think `countDocuments` can be that slow

**Reply by carlcs** [reply]
@r1tsuu I don’t know how `countDocuments()` is used internally with the `find()` API. It could be an issue depending on how often it’s used?

We have now tested with 5 mio docs in the collection. This is the results:

mongoose findOne 26ms
mongoose countDocuments 906ms
payload find 44208ms

```ts
const time0 = Date.now();
const itemMongoose = await payload.db.collections["testitems"].findOne({
  listId: listId,
  isUsed: false,
});
const time1 = Date.now();
console.log("mongoose findOne", time1 - time0);

const count = await payload.db.collections["testitems"]
  .where({
    listId: listId,
    isUsed: false,
  })
  .countDocuments();
const time2 = Date.now();
console.log("mongoose countDocuments", time2 - time1);

const itemPayload = await payload.find({
  collection: "testitems",
  where: { and: [{ listId: { equals: listId } }, { isUsed: { equals: false } }] },
  limit: 1,
});
const time3 = Date.now();
console.log("payload find", time3 - time2);
```

**Reply by r1tsuu** [reply]
It's used internally here (well, actually by https://www.npmjs.com/package/mongoose-paginate-v2 )https://github.com/payloadcms/payload/blob/b546c7b6553f1189881b5466f3bad134e4090bb4/packages/db-mongodb/src/find.ts#L91 that's why output of the `find` function include `totalPages`, `totalDocs`. even with `limit: 1`. I don't think it's used more than once. Maybe try to use `payload.db.collections["testitems"].paginate` directly?

There's a feature request about `payload.findOne` method which would help here definetely https://github.com/payloadcms/payload/discussions/4167

Do you have also any relationships?

**Reply by thijssmudde** [reply]
@carlcs Have you solved any of the performance issues? My API requests are taking too long to load. Directly using mongoose is way faster.

**Reply by carlcs** [reply]
@thijssmudde this was a special case with a very large collection with multiple millions of records, and we did stick to using mongoose directly for our custom endpoints. With more normal size collections this isn’t noticeable that much and we usually stick with just using Payload API.

**Reply by AlessioGr** [reply]
Try upgrading to the latest v3 stable. We've made some huge performance improvements here

**Reply by r1tsuu** [reply]
Also, you should use `pagination: false` (in combination with `limit`) Local API parameter where you don't need count / pages. It removes needing to do count which can improve perf.

**Reply by geminigeek** [reply]
> Also, you should use `pagination: false` (in combination with `limit`) Local API parameter where you don't need count / pages. It removes needing to do count which can improve perf.

why not expose `&pagination=false` in rest api , will get huge performance for any query where expected output is just one , eg (matching unique slug ,  limit 1 )

https://github.com/payloadcms/payload/blob/306b5d230040526acc7645037a3b0f6373c64289/packages/next/src/routes/rest/collections/find.ts#L15

**Reply by Danil0v3s** [reply]
I'm also experiencing something similar with the latest v3 stable, empty database just hopping between collections is slowing everything down

**Comment by vignesh-gupta** [comment]
Hey @jmikrut, I'm sort of agreeing this. 

I'm using version - `3.0.0-beta.70` with nextjs. for me all queries takes quiet some time, not that much but still a bad UX
